src/iop/filmicrgb.c

```diff
@@ -1502,6 +1504,363 @@ static inline void filmic_chroma_v2_v3(const float *const restrict in, float *co
+  dot_product(LMS, matrix, out);
+
+  for_each_channel(c, aligned(out)) out[c] = fmaxf(out[c], 0.f);
+}
+
+static inline void filmic_desaturate_v4(const dt_aligned_pixel_t Ych_original, dt_aligned_pixel_t Ych_final, const float saturation)
+{
+  // Note : Ych is normalized trough the LMS conversion,
+  // meaning c is actually a saturation (saturation ~= chroma / brightness).
+  // So copy-pasting c and h from a different Y is equivalent to
+  // tonemapping with a norm, which is equivalent to doing exposure compensation :
+  // it's saturation-invariant, aka chroma will get increased
+  // if Y is increased, and the other way around.
+  const float chroma_original = Ych_original[1] * Ych_original[0];  // c2
+  float chroma_final = Ych_final[1] * Ych_final[0];                 // c1
+
+  // fit a linear model `chroma = f(y)`:
+  // `chroma = c1 + (yc - y1) * (c2 - c1) / (y2 - y1)`
+  // where `(yc - y1)` is user-defined as `saturation * (y2 - y1)`
+  // so `chroma = c1 + saturation * (c2 - c1)`
+  // when saturation = 0, we stay at the saturation-invariant final chroma
+  // when saturation > 1, we go back towards the initial chroma before tone-mapping
+  // when saturation < 0, we amplify the initial -> final chroma change
+  const float delta_chroma = saturation * (chroma_original - chroma_final);
+
+  const int filmic_brightens = (Ych_final[0] > Ych_original[0]);
+  const int filmic_resat = (chroma_original < chroma_final);
+  const int filmic_desat = (chroma_original > chroma_final);
+  const int user_resat = (saturation > 0.f);
+  const int user_desat = (saturation < 0.f);
+
+  chroma_final = (filmic_brightens && filmic_resat)
+                      ? chroma_original // force original lower sat if brightening
+                  : ((user_resat && filmic_desat) || user_desat)
+                      ? chroma_final + delta_chroma // allow resaturation only if filmic desaturated, allow desat anytime
+                      : chroma_final;
+
+  Ych_final[1] = fmaxf(chroma_final / Ych_final[0], 0.f);
+}
+
+// Pipeline and ICC luminance is CIE Y 1931
+// Kirk Ych/Yrg uses CIE Y 2006
+// 1 CIE Y 1931 = 0.945302269001 CIE Y 2006, so we need to adjust that
+#define CIE_Y_1931_to_CIE_Y_2006(x) 0.945302269001f * x
+
+
+static inline void gamut_check_RGB(dt_aligned_pixel_t RGB_in, dt_aligned_pixel_t Ych_in,
+                                   const dt_colormatrix_t matrix_in, const dt_colormatrix_t matrix_out,
+                                   const float display_black, const float display_white)
+{
+  const int max_iter = 8;
+
+  int iter = 0;
+  float max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  while(max_pix > display_white && iter < max_iter)
+  {
+    // Perform exposure correction to fit back in gamut
+    // This gives us the closest in-gamut pixel at same saturation.
+    dt_aligned_pixel_t RGB_darkened = { 0.f };
+    for_each_channel(c, aligned(RGB_darkened)) RGB_darkened[c] = RGB_in[c] / max_pix;
```

Comment:
Not necessarily. We are in display RGB, so the gamut is bounded in [0; 1] in the 3 dimensions.

Reply:
What about cases where `display_white > 1` i.e. some future HDR displays? After all, the while loop
```c
while(max_pix > display_white && iter < max_iter)
{
  ...
}
```
indeed loops until each RGB component is below `display_white`.

Reply:
Right. The part I'm not sure of is how HDR RGB is encoded. I think it's still encoded between 0 and 1, and then 1 is mapped to 1023 in 10 bits or 4095 in 12 bits. It's just that the meaning of this 1 changes. But then, we have to decide where to normalize to that 1 and it's probably better to do it at the far end, so indeed it might be good to keep the display-referred scaling here (where 1 means reflective white == SDR white, and RGB can go above).

Reply:
Yeah, that makes sense to me.